examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,417 @@
+"""
+Title: Siamese Network with Triplet Loss
+Author: [Hazem Essam](https://twitter.com/hazemessamm) and [Santiago L. Valdarrama](https://twitter.com/svpino)
+Date created: 2021/03/13
+Last modified: 2021/03/22
+Description: Siamese network with tf.data and custom training loop.
+"""
+
+
+"""
+TODO: Still need to work on this.
+
+# Introduction
+
+[Siamese Network](https://en.wikipedia.org/wiki/Siamese_neural_network) is used to solve
+many problems like detecting question duplicates, face recognition by comparing the
+similarity of the inputs by comparing their feature vectors.
+First we need to have a dataset that contains 3 Images, 2 are similar and 1 is different,
+they are called Anchor image, Positive Image and Negative image respectively, we need to
+tell the network that the anchor image and the positive image are similar, we also need
+to tell it that the anchor image and the negative image are NOT similar, we can do that
+by the Triplet Loss Function.
```

Comment:
Suggestion:

- Add a triplet loss reference (Schroff et al., 2015)
- Refactor

```suggestion
First, we need to have a dataset that contains 3 images, of which 2 are similar and 1 is
different. These images are referred to as _anchor_, _positive_, and _negative_ images,
respectively. The neural network will need to know that the anchor and the positive images
are similar, while the anchor and the negative images are dissimilar — this can be done with the help of the [triplet loss](https://www.tensorflow.org/addons/tutorials/losses_triplet)
(you can find out more in the FaceNet paper by [Schroff et al., 2015](https://arxiv.org/pdf/1503.03832.pdf)).
```